i2c: target: eeprom_target: add num-address-bytes parameter

bravl opened the pull request:
Add a parameter to allow the configuration of the number of address bytes used by the I2C master. This allows the driver to expose larger buffer sizes.

Implementation similar to https://github.com/torvalds/linux/blob/master/drivers/i2c/i2c-slave-eeprom.c

Tested with standard linux at24 driver.

bravl committed:
i2c: target: eeprom_target: add address_width parameter

Add a parameter to allow the configuration of the number of address
bytes used by the I2C master. This allows the driver to expose larger
buffer sizes.

Tested with standard linux at24 driver.

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

de-nordic commented:
Isn't this the same as:
https://github.com/zephyrproject-rtos/zephyr/blob/297ad5186d8712a6a261bc0c14a51a1ff73c8011/dts/bindings/mtd/atmel%2Cat2x-base.yaml#L15-L18
for at24 binding?

bravl commented:
> Isn't this the same as:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/297ad5186d8712a6a261bc0c14a51a1ff73c8011/dts/bindings/mtd/atmel%2Cat2x-base.yaml#L15-L18
> 
> 
> for at24 binding?

Yes, that parameter has the same goal as `num-address-bytes` but is related to the at24 driver, this parameter is to allow the target-eeprom driver to support multiple address widths.

I can rename `num-address-bytes` to `address-width` to have it be consistent?

de-nordic commented:
> 
> I can rename `num-address-bytes` to `address-width` to have it be consistent?

I think that maybe the ` address-width` should be moved up from `atmel,at2x-base.yaml` to `zephyr,i2c-target-eeprom.yaml` and the former should become derivative of later (by including it as base).  @henrikbrixandersen will know better.

henrikbrixandersen commented:
> > I can rename `num-address-bytes` to `address-width` to have it be consistent?
> 
> I think that maybe the ` address-width` should be moved up from `atmel,at2x-base.yaml` to `zephyr,i2c-target-eeprom.yaml` and the former should become derivative of later (by including it as base). @henrikbrixandersen will know better.

No, please keep these separate. I think the property should be named the same, though (`address-width`).

bravl commented:
> > > I can rename `num-address-bytes` to `address-width` to have it be consistent?
> > 
> > 
> > I think that maybe the ` address-width` should be moved up from `atmel,at2x-base.yaml` to `zephyr,i2c-target-eeprom.yaml` and the former should become derivative of later (by including it as base). @henrikbrixandersen will know better.
> 
> No, please keep these separate. I think the property should be named the same, though (`address-width`).

Okay. I'll rename the property. Thanks

jacmet commented on dts/bindings/mtd/zephyr,i2c-target-eeprom.yaml:
```diff
@@ -5,3 +5,11 @@ description: Zephyr I2C target EEPROM
+    enum: [8, 16]
+    description: |
+      Number of address bits used to address the EEPROM. If not specified
+      the EEPROM is assumed to have a 8-bit address.
```
We have both size (from eeprom-base.yaml) and address-width, and they should (normally) be in sync. How about figuring out the number of address bytes from the size (E.G. 8bit for <= 256, else 16bit)?

bravl replied:
@henrikbrixandersen what would be the preferred approach?
I agree with @jacmet it would be "cleaner" if the address width is linked to the actual size. However, making it dynamic means it would break existing use where the size is set larger than 256. Even though right now, it's not possible to address larger sizes. 

I saw that some overlays uses in the eeprom target tests actually have the size set to 1024. These overlays would then also need to be updated.

henrikbrixandersen replied:
`address-width` and `size` are not necessarily tightly coupled in "real" AT24 EEPROMs. It depends on the family of the EEPROM. 

I don't think we should tightly-couple them here either.

henrikbrixandersen replied:
> you mean things like 24c04 where the upper address bits "spill" into the I2C address? Indeed, but the driver does not support that today (and some I2C drivers only support 1 or very few target addresses).

Correct, but if we introduce the automatic derivation of one from the other, it will make it more difficult to add support for something like the above later on.

A use-case could be to use Zephyr for emulating (part of) a "real" EEPROM, where the user is not in control of the I2C "master" and thus needs to mimic the address-width of the device, but perhaps not the size. This can be needed due to obsolescence of the original part, redesigns, ...

> Even if not tightly coupled, they are certainly related - E.G. it doesn't make any sense to use a size that cannot be represented in the address byte(s)

As long as we error out if attempting to write past the actual buffer size, I don't see a need for this restriction.

bravl replied:
Okay, then we can leave it as implemented where size and address-width are not directly linked to each other.

jacmet replied:
OK, fine by me. The current behavior is that the address provided over I2C is used mod size, so it wraps around if "too many" I2C address bits are available. Do you still want to allow bigger buffer sizes than what can be addressed over I2C, E.G. wasted inaccessible memory like the current tests with 1 address byte and 1024 bytes?

henrikbrixandersen replied:
I'd suggest adding a build assert checking for too big buffers compared to `address-width`.

jacmet replied:
@henrikbrixandersen  OK. That presumably alsa means updating the board overlays in the tests/drivers/i2c/i2c_target_api test that do not set address-width (E.G. = 1 byte) and use a 1024 byte buffer today to use a 256 byte butter?

henrikbrixandersen replied:
Yes, please. Otherwise they would fail this build assertion.

bravl replied:
Ok! I'll add the build define. W.r.t the tests, I'll move some tests to size 256 and from some I'll add address-width 16 in order to test both case.

bravl replied:
Added build-time asserts and fixed the sizes to 256 in the tests.

teburd commented:
Re-assigned to @henrikbrixandersen as he's most familiar with eeprom related code

bravl committed:
i2c: target: eeprom_target: add addressable size assert

Add build-time assert to check if the size defined is actually addressable.
i.e: if address-width is set to 8, the maximum addressable size is 256
bytes.

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

bravl committed:
tests: i2c: i2c_target_api: fix test before size build assert addition

Fix the tests before a build time check is added to validate the
addressable size. Most tests have an eeprom size of 1024 defined but width
8-bit addressing, the majority of the eeprom cannot be addressed

Signed-off-by: Bram Vlerick <bram.vlerick@openpixelsystems.org>

jacmet commented on tests/drivers/i2c/i2c_target_api/boards/b_u585i_iot02a.overlay:
```diff
@@ -14,14 +14,14 @@
 		compatible = "zephyr,i2c-target-eeprom";
 		reg = <0x54>;
-		size = <1024>;
+		size = <256>;
```
NIT: This change should be done before the build assert for bisectability

jacmet commented on drivers/i2c/target/eeprom_target.c:
```diff
@@ -259,6 +259,10 @@ static int i2c_eeprom_target_init(const struct device *dev)
 									\
+	BUILD_ASSERT(DT_INST_PROP(inst, size) <=			\
+			(1 << DT_INST_PROP_OR(inst, address_width, 8)), \
+			"size must be < than 2^address_width");		\
```
NIT: <=, not <

bravl replied:
Fixed!

bravl commented on tests/drivers/i2c/i2c_target_api/boards/b_u585i_iot02a.overlay:
```diff
@@ -10,6 +10,8 @@
  * Short Pin PB9 to PH5, and PB8 to PH4, for the test to pass.
  */
 
+/delete-node/ &eeprom0;
```
Indeed but the build of the samples fails without this fix causing the PR to be 'rejected' based on the merge criteria.

bravl replied:
I could move it to a separate PR and keep this one open until that fix is merged.

bravl replied:
Dropped fix and created separate PR: https://github.com/zephyrproject-rtos/zephyr/pull/72290

bravl replied:
Rebased and builds cleanly now after merge of PR #72290

bravl commented:
:wave: @henrikbrixandersen any further feedback on the proposed changes?

jacmet reviewed (approved):
LGTM, thanks
